designs/resource-properties.md

```diff
@@ -0,0 +1,490 @@
+
+## Alternative approaches
+
+### (`@notProperty` alternative) Add `ignoredProperties` to resource
+
+Rather than define the `@notProperty` trait on members, it could be defined on
+the resource directly:
+
+```
+resource Config {
+    identifiers: {
+        configId: ConfigId
+    }
+    properties: {
+        configData: ConfigData
+        name: ConfigName
+        tags: TagList
+        configArn: ConfigArg
+        configType: ConfigType
+    }
+    ignoredProperties: {
+        token: String
+    }
+    create: CreateConfig
+}
+```
+
+This could be made more succinct with mixins:
+
+```
+@mixin
+resource BaseResourceMixin {
+    ignoredProperties: {
+        token: String
+    }
+}
+
+resource Config with [BaseResourceMixin] {
+    identifiers: {
+        configId: ConfigId
+    }
+    properties: {
+        configData: ConfigData
+        name: ConfigName
+        tags: TagList
+        configArn: ConfigArg
+        configType: ConfigType
+    }
+    create: CreateConfig
+}
+```
+
+Pros:
+
+* Remove the need to add property traits to members spread across the resource.
+
+Cons:
+
+* The resource shape becomes verbose. It leaks details of create and instance
+  operation input and output members that need to be filtered out.
+* More difficult to process ignoredProperties when model undergoes transformations
+  that may filter the operations on a service or resource. Operation inclusion
+  or exclusion would impact the contents of ignoredProperties.
+
+### (`@nestedProperties` alternative): Associating top level resource properties to structures with `@associatedResource`
+
+In order to support existing services migration to resource properties we could
+introduce a new trait named `@associatedResource` to associate a structure shape
+with the top level properties of a specific resource instead of a top level
+lifecycle operation input or output. This trait can only be allowed to structure
+shapes one below the input or output of a resource lifecycle operation.
+
+```
+resource Pipeline {
+...
+     properties: {
+        name: String
+        foo: Baz
+        rank: Integer
+    },
+    create: CreatePipeline
+    ...
+}
+
+operation CreatePipeline {
+    input := CreatePipelineInput
+}
+
+structure CreatePipelineInput {
+     pipeline: PipelineDescription
+}
+
+@associatedResource("Pipeline")
+structure PipelineDescription {
+     name: String
+    foo: Baz
+    @property(name:"rank")
+    ranking: Integer
+ }
+```
+
+#### `@associatedResource` trait definition
+
+The `@associatedResource` trait is defined in Smithy’s prelude as:
+
+```
+@trait(selector: "structure") //TODO: How to force off of top level input/output
+string associatedResource
+```
+
+Pros:
+
+* Directly associates the resource with the shape to find the properties
+  regardless of how deeply nested it is.
+
+Cons:
+
+* Should not allow nesting of properties beyond 1 level deep.
+* Service author may incidentally associate more than one resource in a input
```

Comment:
it also adds a lot of complexity around what happens if there are top-level members that qualify as a resource property, but also nested in the resource, there are other properties... what if those properties conflict?

There's also the issue of readability -- in order to understand the resource properties, you can to poke around at all the members.